## Gather

In [83]:
import pandas as pd
import numpy as np

In [84]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [85]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [86]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [87]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [88]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [89]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [90]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [91]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [92]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [93]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [94]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [95]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [96]:
patients.sample(5)

patient_id assigned_sex given_name    surname               address  \
5             6         male     Rafael      Costa    1140 Willis Avenue   
347         348       female      Urška   Aličajić    3781 Hamill Avenue   
81           82         male      André     Nordin  836 Dovetail Estates   
335         336       female      Lixue      Hsueh   1540 Overlook Drive   
260         261       female      Emily  Henriksen     4236 Beech Street   

               city    state  zip_code        country  \
5     Daytona Beach  Florida   32114.0  United States   
347       San Diego       CA   92109.0  United States   
81             Gate       OK   73844.0  United States   
335  Crawfordsville       IN   47933.0  United States   
260         Antioch       CA   94509.0  United States   

                                      contact   birthdate  weight  height  \
5    386-334-5237RafaelCardosoCosta@gustr.com   8/31/1931   183.9      70   
347       858-273-2043UrskaAlicajic@rhyta.com   6/27/1952   183.3      63   
81           580-934-1141AndreNordin@cuvox.de  11/12/1940   209.4      69   
335         765-359-0147LixueHsueh@dayrep.com   3/29/1925   102.7      59   
260   EmilyNHenriksen@armyspy.com925-757-6139    7/8/1957   206.4      66   

      bmi  
5    26.4  
347  32.5  
81   30.9  
335  20.7  
260  33.3

In [97]:
patients.surname.value_counts()

Doe               6
Jakobsen          3
Taylor            3
Lund              2
Tucker            2
Johnson           2
Tạ                2
Cindrić           2
Schiavone         2
Parker            2
Hueber            2
Lâm               2
Woźniak           2
Gersten           2
Dratchev          2
Grímsdóttir       2
Kowalczyk         2
Collins           2
Aranda            2
Batukayev         2
Silva             2
Souza             2
Lương             2
Liễu              2
Correia           2
Kadyrov           2
Cabrera           2
Ogochukwu         2
Berg              2
Nilsen            2
                 ..
Azuma             1
Wellish           1
Guðbrandsson      1
Tomaszewski       1
Onio              1
Frederiksen       1
Alanis            1
Afanasyeva        1
Sakai             1
Eldarkhanov       1
Eidem             1
Carreiro          1
Yonatan           1
Ménard            1
Chidi             1
Halldórsdóttir    1
Vaara             1
Petersen          1
O'Brian           1


In [98]:
patients.address.value_counts()

123 Main Street               6
648 Old Dear Lane             2
2476 Fulton Street            2
2778 North Avenue             2
3558 Longview Avenue          1
212 Tibbs Avenue              1
4643 Reeves Street            1
954 Summit Park Avenue        1
3391 Marcus Street            1
3450 Southern Street          1
3303 Anmoore Road             1
1054 Zappia Drive             1
3417 Bassel Street            1
2549 Pearlman Avenue          1
4262 Heron Way                1
4912 Hart Country Lane        1
707 Gateway Avenue            1
142 Broad Street              1
1231 Grim Avenue              1
3141 Brentwood Drive          1
945 Maple Avenue              1
4992 Moonlight Drive          1
4556 Oakdale Avenue           1
57 Norman Street              1
547 Weekley Street            1
2458 Broadway Avenue          1
1813 Lindale Avenue           1
441 Tibbs Avenue              1
1262 Deans Lane               1
1116 Dog Hill Lane            1
                             ..
2418 Smi

In [99]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [100]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [101]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [102]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [103]:
sum(treatments.auralin.isnull())

0

In [104]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [105]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
*concatenate the missing `treatments` records which are stored in a file named `treatments_cut.csv` (having 70 records) with the original `treatments` dataframe*

In [106]:
missing_treatments = pd.read_csv('treatments_cut.csv')
missing_treatments.shape

(70, 7)

##### Code

In [107]:
# Using the concat function to join the rows in both treatment and missing_treatments dataframes

treatments_clean = pd.concat([treatments_clean, missing_treatments], axis=0, ignore_index=True)

##### Test

In [108]:
# Your testing code here
treatments_clean.shape

(350, 7)

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
*replacing the `hba1c_change` column with a another having correct and complete values*

In [109]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

##### Code

In [110]:
# Creating a new hba1c_change column with correct and complete values
treatments_clean['hba1c_change'] =  treatments_clean['hba1c_start'] - treatments_clean['hba1c_end']

##### Test

In [111]:
# Your testing code here
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

In [112]:
treatments_clean.hba1c_change.isnull().sum()

0

### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

In [113]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

##### Code

In [114]:
# Your cleaning code here
patients_clean['email'] = patients_clean['contact'].str.extract('([a-zA-Z]+@[a-z]+\.{1}[a-z]{0,3})').astype(str)

##### Test

In [115]:
# Your testing code here
patients_clean[['contact', 'email']].sample(10)

contact  \
454   BiancaCarvalhoAzevedo@einrot.com1 606 589 2819   
140           217-421-0935JohnTeichelmann@dayrep.com   
274                740-552-2444NoraNyborg@einrot.com   
33                815-270-7770HabteKifle@fleckens.hu   
174           509-466-2907MichelleDTansey@dayrep.com   
338  LoaHrafnsdottir@jourrapide.com+1 (985) 335-8662   
475           MalikVaneker@superrito.com650-962-7179   
176            MakdaAbraham@einrot.com1 508 857 0477   
318             BenoitBonami@gustr.com1 718 954 8136   
303            JoeVEdwards@armyspy.com1 313 589 1746   

                                email  
454  BiancaCarvalhoAzevedo@einrot.com  
140        JohnTeichelmann@dayrep.com  
274             NoraNyborg@einrot.com  
33             HabteKifle@fleckens.hu  
174        MichelleDTansey@dayrep.com  
338    LoaHrafnsdottir@jourrapide.com  
475        MalikVaneker@superrito.com  
176           MakdaAbraham@einrot.com  
318            BenoitBonami@gustr.com  
303           JoeVEdwards@armyspy.com

In [116]:
patients_clean['phone_number'] = patients_clean['contact'].str.extract('([^a-z]{0,3}\+?1?[\s(]?[0-9]{3}[\)]?[\s-]?[0-9]{3}[\s-]?[0-9]{4})').astype(str)

In [117]:
patients_clean[['phone_number','email']].sample(10)

phone_number                               email
47     404-289-1766           AnaSousaCorreia@rhyta.com
415    909-861-4356              MgbekeEkwueme@cuvox.de
427    325-513-8964            TeclaOnio@jourrapide.com
357    937-592-9410         ChiemelaTobeolisa@rhyta.com
65     765-379-3885  NicolasSouzaFerreira@superrito.com
318  1 718 954 8136              BenoitBonami@gustr.com
109    601-699-4153     StephenFMayberry@jourrapide.com
92     609-914-8473            MichaelSmith@armyspy.com
464    205-467-9279           MeazaBrhane@superrito.com
407    405-470-5088              TeganJohnson@gustr.com

In [118]:
patients_clean = patients_clean.drop(['contact'], axis=1)

In [119]:
patients_clean = patients_clean.reindex(['patient_id', 'assigned_sex', 'given_name', 'surname', 'address',
       'city', 'state', 'zip_code', 'country', 'email', 'phone_number', 'birthdate', 'weight', 'height',
       'bmi'], axis=1)

In [120]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                       email       phone_number  birthdate  weight  height  \
0   ZoeWellish@superrito.com       951-719-9170  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de  +1 (217) 569-3204   4/3/1967   118.8      66   
2       JaeMDebord@gustr.com       402-363-6804  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com  +1 (732) 636-8246  7/26/1951   220.9      70   
4        TimNeudorf@cuvox.de       334-515-7487  2/18/1928   192.3      27   

    bmi  
0  19.6  
1  19.2  
2  24.8  
3  31.7  
4  26.1

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

In [121]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

##### Code

In [122]:
# Your cleaning code here
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean.head(2)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1     elliot  richardson         7.56       7.09          0.47   auralin   

        dose  
0  41u - 48u  
1          -

In [123]:
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean

given_name        surname  hba1c_start  hba1c_end  hba1c_change  \
0        veronika       jindrová         7.63       7.20          0.43   
3            skye     gormanston         7.97       7.62          0.35   
6          sophia         haugen         7.65       7.27          0.38   
7           eddie         archer         7.89       7.55          0.34   
9            asia        woźniak         7.76       7.37          0.39   
10         joseph            day         7.70       7.19          0.51   
12        roxanne      andreyeva         9.54       9.14          0.40   
14         simone   baumgaertner         7.74       7.30          0.44   
15           enco         žibrik         7.78       7.34          0.44   
16        camilla       zaitseva         7.53       7.13          0.40   
21          tekla        walczak         7.61       7.29          0.32   
22    brancaleone          russo         8.61       8.18          0.43   
24           isac           berg         9.68       9.29          0.39   
28        clinton         miller         7.79       7.40          0.39   
29         eugene        mironov         7.81       7.48          0.33   
30    szilveszter          totth         7.70       7.38          0.32   
34      alexander      mathiesen         7.96       7.55          0.41   
36            chỉ            lâm         7.68       7.24          0.44   
37      władysław      wieczorek         7.92       7.47          0.45   
38       kristján        ingason         7.92       7.57          0.35   
42         marija       grubišić         7.53       7.15          0.38   
44          sauli     koivuniemi         7.67       7.37          0.30   
46        mariana          souza         7.86       7.51          0.35   
51     kristoffer      martinsen         9.18       8.64          0.54   
52             mỹ          quynh         7.61       7.16          0.45   
54           oles        zhdanov         7.52       7.11          0.41   
55        triana.       terrazas         7.71       7.34          0.37   
58         gabryŝ    tomaszewski         7.87       7.47          0.40   
59      leixandre         alanis         7.74       7.32          0.42   
60   onyekachukwu         obinna         7.58       7.12          0.46   
..            ...            ...          ...        ...           ...   
638          jane        citizen         7.98       7.60          0.38   
639        angela     lavrentyev         7.61       7.14          0.47   
640        edelma    villalpando         7.99       7.56          0.43   
642        annika          vaara         7.73       7.34          0.39   
644         chiho           higa         7.71       7.30          0.41   
652     beatrycze        woźniak         7.54       7.17          0.37   
653      miłosław     wiśniewski         7.51       7.08          0.43   
656       firenze          fodor         7.89       7.55          0.34   
657           zoe        wellish         7.71       7.30          0.41   
658           una  traustadóttir         8.00       7.50          0.50   
659         luboš          pecha         7.79       7.45          0.34   
663         meaza         brhane         7.70       7.36          0.34   
664         adlan       shishani         7.84       7.37          0.47   
667         sofia      hermansen         8.90       8.57          0.33   
668         guðni      heimisson         7.64       7.24          0.40   
670       eufemio        rosario         7.54       7.26          0.28   
671      dalmacia         madrid         7.67       7.21          0.46   
672         daimy          tromp         9.41       8.94          0.47   
674        jeremy        montagu         7.68       7.36          0.32   
678       nebechi      ekechukwu         7.78       7.39          0.39   
679       satsita      batukayev         7.63       7.25          0.38   
680       timothy         cotton         7.92       7.52          0.40   
681       b

##### Test

In [124]:
# Your testing code here
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

        dose  
0  41u - 48u  
3  33u - 36u  
6  37u - 42u  
7  31u - 38u  
9  30u - 36u

#### Adverse reaction should be part of the `treatments` table

##### Define
merge (left join) the treatments_clean and adverver_reaction_clean tables together

In [125]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

        dose  
0  41u - 48u  
3  33u - 36u  
6  37u - 42u  
7  31u - 38u  
9  30u - 36u

In [126]:
adverse_reactions_clean.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

##### Code

In [127]:
# Your cleaning code here
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on = ['given_name','surname']
                            , how='left')

##### Test

In [128]:
# Your testing code here
treatments_clean.head(6)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1       skye  gormanston         7.97       7.62          0.35   auralin   
2     sophia      haugen         7.65       7.27          0.38   auralin   
3      eddie      archer         7.89       7.55          0.34   auralin   
4       asia     woźniak         7.76       7.37          0.39   auralin   
5     joseph         day         7.70       7.19          0.51   auralin   

        dose adverse_reaction  
0  41u - 48u              NaN  
1  33u - 36u              NaN  
2  37u - 42u              NaN  
3  31u - 38u              NaN  
4  30u - 36u              NaN  
5  29u - 36u     hypoglycemia

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

##### Code

In [129]:
# Your cleaning code here
id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

/opt/conda/lib/python3.6/site-packages/pandas/core/generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


##### Test

In [130]:
# Your testing code here
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment       dose adverse_reaction  \
0         7.63       7.20          0.43   auralin  41u - 48u              NaN   
1         7.97       7.62          0.35   auralin  33u - 36u              NaN   
2         7.65       7.27          0.38   auralin  37u - 42u              NaN   
3         7.89       7.55          0.34   auralin  31u - 38u              NaN   
4         7.76       7.37          0.39   auralin  30u - 36u              NaN   

   patient_id  
0         225  
1         242  
2         345  
3         276  
4          15

In [131]:
treatments_clean = treatments_clean.reindex([ 'patient_id','hba1c_start', 'hba1c_end', 'hba1c_change', 'treatment', 'dose',
       'adverse_reaction'], axis = 1)
treatments_clean.head()

patient_id  hba1c_start  hba1c_end  hba1c_change treatment       dose  \
0         225         7.63       7.20          0.43   auralin  41u - 48u   
1         242         7.97       7.62          0.35   auralin  33u - 36u   
2         345         7.65       7.27          0.38   auralin  37u - 42u   
3         276         7.89       7.55          0.34   auralin  31u - 38u   
4          15         7.76       7.37          0.39   auralin  30u - 36u   

  adverse_reaction  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
change the datatype of the zipcode to strings, ensure all zip codes are five digits long

##### Code

In [132]:
import numpy as np

In [133]:
# Your cleaning code here
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar='0')


##### Test

In [134]:
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)

In [135]:
# Your testing code here
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  \
0  Rancho California  California    92390  United States   
1          Armstrong    Illinois    61812  United States   
2               York    Nebraska    68467  United States   
3         Woodbridge          NJ    07095  United States   
4             Dothan          AL    36303  United States   

                       email       phone_number  birthdate  weight  height  \
0   ZoeWellish@superrito.com       951-719-9170  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de  +1 (217) 569-3204   4/3/1967   118.8      66   
2       JaeMDebord@gustr.com       402-363-6804  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com  +1 (732) 636-8246  7/26/1951   220.9      70   
4        TimNeudorf@cuvox.de       334-515-7487  2/18/1928   192.3      27   

    bmi  
0  19.6  
1  19.2  
2  24.8  
3  31.7  
4  26.1

In [136]:
patients.head(10)

patient_id assigned_sex given_name     surname                     address  \
0           1       female        Zoe     Wellish        576 Brown Bear Drive   
1           2       female     Pamela        Hill   2370 University Hill Road   
2           3         male        Jae      Debord       1493 Poling Farm Road   
3           4         male       Liêm        Phan         2335 Webster Street   
4           5         male        Tim     Neudorf        1428 Turkey Pen Lane   
5           6         male     Rafael       Costa          1140 Willis Avenue   
6           7       female       Mary       Adams            3145 Sheila Lane   
7           8       female     Xiuxiu       Chang  2687 Black Oak Hollow Road   
8           9         male      Dsvid  Gustafsson          1790 Nutter Street   
9          10       female     Sophie     Cabrera           3303 Anmoore Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   
5      Daytona Beach     Florida   32114.0  United States   
6            Burbank          NV   84728.0  United States   
7        Morgan Hill          CA   95037.0  United States   
8        Kansas City          MO   64105.0  United States   
9           New York    New York   10011.0  United States   

                                         contact   birthdate  weight  height  \
0           951-719-9170ZoeWellish@superrito.com   7/10/1976   121.7      66   
1          PamelaSHill@cuvox.de+1 (217) 569-3204    4/3/1967   118.8      66   
2               402-363-6804JaeMDebord@gustr.com   2/19/1980   177.8      71   
3     PhanBaLiem@jourrapide.com+1 (732) 636-8246   7/26/1951   220.9      70   
4                334-515-7487TimNeudorf@cuvox.de   2/18/1928   192.3      27   
5       386-334-5237RafaelCardosoCosta@gustr.com   8/31/1931   183.9      70   
6              775-533-5933MaryBAdams@einrot.com  11/19/1969   146.3      65   
7           XiuxiuChang@einrot.com1 408 778 3236   8/13/1958   158.0      60   
8        816-265-9578DavidGustafsson@armyspy.com    3/6/1937   163.9      66   
9  SophieCabreraIbarra@teleworm.us1 718 795 9124   12/3/1930   194.7      64   

    bmi  
0  19.6  
1  19.2  
2  24.8  
3  31.7  
4  26.1  
5  26.4  
6  24.3  
7  30.9  
8  26.5  
9  33.4

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Change Neudorf's height to 72*

##### Code

In [137]:
# Your cleaning code here
patients_clean.iloc[4,13] = 72

##### Test

In [138]:
# Your testing code here

patients_clean[patients_clean.surname == "Neudorf"]

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country                email  phone_number  birthdate  \
4    AL    36303  United States  TimNeudorf@cuvox.de  334-515-7487  2/18/1928   

   weight  height   bmi  
4   192.3      72  26.1

#### Full state names sometimes, abbreviations other times

##### Define
*Change all the states names to short forms*

In [139]:
patients_clean.state.unique()

array(['California', 'Illinois', 'Nebraska', 'NJ', 'AL', 'Florida', 'NV',
       'CA', 'MO', 'New York', 'MI', 'TN', 'VA', 'OK', 'GA', 'MT', 'MA',
       'NY', 'NM', 'IL', 'LA', 'PA', 'CO', 'ME', 'WI', 'SD', 'MN', 'FL',
       'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD', 'AZ', 'TX',
       'NE', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH', 'OR',
       nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)

##### Code

In [140]:
# Your cleaning code here
patients_clean.replace({'state':['California','Illinois','Nebraska','Florida','New York']},
                       {'state':['CA','IL','NE','FL','NY']}, inplace=True)

##### Test

In [141]:
# Your testing code here
patients_clean.state.unique()

array(['CA', 'IL', 'NE', 'NJ', 'AL', 'FL', 'NV', 'MO', 'NY', 'MI', 'TN',
       'VA', 'OK', 'GA', 'MT', 'MA', 'NM', 'LA', 'PA', 'CO', 'ME', 'WI',
       'SD', 'MN', 'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD',
       'AZ', 'TX', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH',
       'OR', nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)

#### Dsvid Gustafsson

##### Define
*Replace given names of 'Dsvid' with 'David'*

##### Code

In [142]:
# Your cleaning code here
patients_clean.given_name = patients_clean.given_name.replace('Dsvid','David')

##### Test

In [143]:
# Your testing code here
patients_clean[patients_clean.surname == "Gustafsson"]

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country                        email  \
8  Kansas City    MO    64105  United States  DavidGustafsson@armyspy.com   

   phone_number birthdate  weight  height   bmi  
8  816-265-9578  3/6/1937   163.9      66  26.5

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Convert assigned sex and state to categorical data types. Zip code data type was already addressed above. Convert birthdate to datetime data type. Strip the letter 'u' in start dose and end dose and convert those columns to data type integer.*

##### Code

In [144]:
treatments_clean.head(2)

patient_id  hba1c_start  hba1c_end  hba1c_change treatment       dose  \
0         225         7.63       7.20          0.43   auralin  41u - 48u   
1         242         7.97       7.62          0.35   auralin  33u - 36u   

  adverse_reaction  
0              NaN  
1              NaN

In [145]:
treatments_clean['dose_start'] = treatments_clean.dose.apply(lambda x: x.split('-')[1])
treatments_clean['dose_end'] = treatments_clean.dose.apply(lambda x: x.split('-')[0])
treatments_clean = treatments_clean.drop('dose', axis=1)
treatments_clean.head()

patient_id  hba1c_start  hba1c_end  hba1c_change treatment  \
0         225         7.63       7.20          0.43   auralin   
1         242         7.97       7.62          0.35   auralin   
2         345         7.65       7.27          0.38   auralin   
3         276         7.89       7.55          0.34   auralin   
4          15         7.76       7.37          0.39   auralin   

  adverse_reaction dose_start dose_end  
0              NaN        48u     41u   
1              NaN        36u     33u   
2              NaN        42u     37u   
3              NaN        38u     31u   
4              NaN        36u     30u

In [146]:
treatments_clean['dose_start'] = treatments_clean['dose_start'].apply(lambda x: x.strip())
treatments_clean['dose_end'] = treatments_clean['dose_end'].apply(lambda x: x.strip())

In [147]:
# Your cleaning code here
# assigned sex and state columns to category
patients_clean[['assigned_sex','state']] = patients_clean[['assigned_sex','state']].astype('category')

# birthday to datetime
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

#remove the 'u' in the values of the dose column
treatments_clean.dose_start = treatments_clean.dose_start.str.strip('u').astype(int)
treatments_clean.dose_end = treatments_clean.dose_end.str.strip('u').astype(int)

In [148]:
list(treatments_clean)

['patient_id',
 'hba1c_start',
 'hba1c_end',
 'hba1c_change',
 'treatment',
 'adverse_reaction',
 'dose_start',
 'dose_end']

In [149]:
treatments_clean = treatments_clean.reindex(['patient_id','hba1c_start','hba1c_end','hba1c_change',
                                             'treatment','dose_start','dose_end','adverse_reaction'], axis=1)
treatments_clean.head()

patient_id  hba1c_start  hba1c_end  hba1c_change treatment  dose_start  \
0         225         7.63       7.20          0.43   auralin          48   
1         242         7.97       7.62          0.35   auralin          36   
2         345         7.65       7.27          0.38   auralin          42   
3         276         7.89       7.55          0.34   auralin          38   
4          15         7.76       7.37          0.39   auralin          36   

   dose_end adverse_reaction  
0        41              NaN  
1        33              NaN  
2        37              NaN  
3        31              NaN  
4        30              NaN

##### Test

In [150]:
# Your testing code here
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        491 non-null object
country         491 non-null object
email           503 non-null object
phone_number    503 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.9+ KB


In [151]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
patient_id          349 non-null int64
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
treatment           349 non-null object
dose_start          349 non-null int64
dose_end            349 non-null int64
adverse_reaction    35 non-null object
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


#### Multiple phone number formats

##### Define
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

##### Code

In [152]:
patients_clean.phone_number.head()

0         951-719-9170
1    +1 (217) 569-3204
2         402-363-6804
3    +1 (732) 636-8246
4         334-515-7487
Name: phone_number, dtype: object

In [153]:
# Your cleaning code here
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

##### Test

In [154]:
# Your testing code here
patients_clean.phone_number.sample(5)

357    19375929410
427    13255138964
203    16364426946
177    13233274721
414    19375261152
Name: phone_number, dtype: object

#### Default John Doe data

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

##### Code

In [155]:
patients_clean[patients_clean.surname == 'Doe']

patient_id assigned_sex given_name surname          address      city  \
215         216         male       John     Doe  123 Main Street  New York   
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state zip_code        country              email phone_number  birthdate  \
215    NY    12345  United States  johndoe@email.com  11234567890 1975-01-01   
229    NY    12345  United States  johndoe@email.com  11234567890 1975-01-01   
237    NY    12345  United States  johndoe@email.com  11234567890 1975-01-01   
244    NY    12345  United States  johndoe@email.com  11234567890 1975-01-01   
251    NY    12345  United States  johndoe@email.com  11234567890 1975-01-01   
277    NY    12345  United States  johndoe@email.com  11234567890 1975-01-01   

     weight  height   bmi  
215   180.0      72  24.4  
229   180.0      72  24.4  
237   180.0      72  24.4  
244   180.0      72  24.4  
251   180.0      72  24.4  
277   180.0      72  24.4

In [156]:
# Your cleaning code here
# Since all the patients with the surname 'Doe' have first name as John, we can filter it out

patients_clean = patients_clean[patients_clean.surname != 'Doe']


##### Test

In [157]:
# Your testing code here
'Doe' in patients_clean.surname

False

#### Multiple records for Jakobsen, Gersten, Taylor

In [158]:
patients_clean[((patients_clean.surname == 'Taylor'))]

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
282         283       female      Sandy  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country                       email phone_number  \
131    WV    25962  United States    SandraCTaylor@dayrep.com  13044382648   
282    WV    25962  United States    SandraCTaylor@dayrep.com  13044382648   
426    FL    33179  United States  RogelioJTaylor@teleworm.us  13054346299   

     birthdate  weight  height   bmi  
131 1960-10-23   206.1      64  35.4  
282 1960-10-23   206.1      64  35.4  
426 1992-09-02   186.6      69  27.6

##### Define
*Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the patients table and keep the original record with correct fullnames*

In [159]:
patients_clean[(patients_clean.address.duplicated() & patients_clean.address.notnull())]

patient_id assigned_sex given_name   surname             address  \
29           30         male       Jake  Jakobsen   648 Old Dear Lane   
282         283       female      Sandy    Taylor  2476 Fulton Street   
502         503         male        Pat   Gersten   2778 North Avenue   

            city state zip_code        country                      email  \
29   Port Jervis    NY    12771  United States  JakobCJakobsen@einrot.com   
282     Rainelle    WV    25962  United States   SandraCTaylor@dayrep.com   
502         Burr    NE    68324  United States   PatrickGersten@rhyta.com   

    phone_number  birthdate  weight  height   bmi  
29   18458587707 1985-08-01   155.8      67  24.4  
282  13044382648 1960-10-23   206.1      64  35.4  
502  14028484923 1954-05-03   138.2      71  19.3

##### Code

In [167]:
# Your cleaning code here
patients_clean = patients_clean[~(patients_clean.address.duplicated() & patients_clean.address.notnull())]

##### Test

In [168]:
# Your testing code here
patients_clean[patients_clean.surname.isin(['Jakobsen', 'Gersten', 'Taylor'])]

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
97           98         male    Patrick   Gersten   2778 North Avenue   
131         132       female     Sandra    Taylor  2476 Fulton Street   
426         427         male    Rogelio    Taylor  4064 Marigold Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country                         email  \
24   Port Jervis    NY    12771  United States     JakobCJakobsen@einrot.com   
97          Burr    NE    68324  United States      PatrickGersten@rhyta.com   
131     Rainelle    WV    25962  United States      SandraCTaylor@dayrep.com   
426        Miami    FL    33179  United States    RogelioJTaylor@teleworm.us   
432      Houston    TX    77020  United States  KarenJakobsen@jourrapide.com   

    phone_number  birthdate  weight  height   bmi  
24   18458587707 1985-08-01   155.8      67  24.4  
97   14028484923 1954-05-03   138.2      71  19.3  
131  13044382648 1960-10-23   206.1      64  35.4  
426  13054346299 1992-09-02   186.6      69  27.6  
432  19792030438 1962-11-25   185.2      67  29.0

#### kgs instead of lbs for Zaitseva weight

##### Define
*change Zaitseva's weight to lbs*

##### Code

In [170]:
# Your cleaning code here
kg_weight = patients_clean.weight.min()
kg_weight

48.799999999999997

In [180]:
patients_clean[patients_clean.surname == 'Zaitseva']

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country                          email  \
210  Wooster    OH    44691  United States  CamillaZaitseva@superrito.com   

    phone_number  birthdate  weight  height   bmi  
210  13302022145 1938-11-26    48.8      63  19.1

In [182]:
patients_clean.loc[210, 'weight'] = kg_weight * 2.20462

##### Test

In [ ]:
# Your testing code here